### PR TITLE
feat: add rules_of_hooks rule

### PR DESCRIPTION
**Pull request by marvinhagemeister**
This PR implements the `rules-of-hooks` eslint rule, see https://react.dev/reference/rules/rules-of-hooks . It ensures that hooks are only used inside components and that they are called unconditionally.

### CODE REVIEW

File: docs/rules/rules_of_hooks.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+Ensure that (P)React hooks are only called inside component functions and not
```
Comment (bartlomieju) [inline docs/rules/rules_of_hooks.md:1]:
```suggestion
Ensure that React/Preact hooks are only called inside component functions and not
```

File: docs/rules/rules_of_hooks.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+called conditionally. A hook is a function that starts with `use`
```
Comment (bartlomieju) [inline docs/rules/rules_of_hooks.md:2]:
```suggestion
called conditionally. A hook is a function that starts with `use` prefix.
```

File: docs/rules/rules_of_hooks.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+```tsx
+function Component() {
+  const [count, setCount] = useState(0);
+  // ...
+}
+
+function useCustomHook() {
+  const [count, setCount] = useState(0);
+  // ...
+}
```
Comment (bartlomieju) [inline docs/rules/rules_of_hooks.md:73]: Could you update this one to show all "bad" examples written in a correct form?

File: src/rules/rules_of_hooks.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+use deno_ast::SourceRanged;
+
+#[derive(Debug)]
+pub struct RulesOfHooks;
+
+const CODE: &str = "rules-of-hooks";
+
+impl LintRule for RulesOfHooks {
+  fn tags(&self) -> &'static [&'static str] {
+    &["react", "jsx", "fresh"]
```
Comment (bartlomieju) [inline src/rules/rules_of_hooks.rs:16]:
```suggestion
    &["recommended", "react", "jsx", "fresh"]
```

File: src/rules/rules_of_hooks.rs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+fn is_hook_or_component_name(name: &str) -> bool {
+  if name.starts_with("use") {
+    return true;
+  }
+
+  if let Some(ch) = name.chars().next() {
+    return ch.is_uppercase();
+  }
```
Comment (bartlomieju) [inline src/rules/rules_of_hooks.rs:276]: Where is this hit? Could you add a small unit test for this function?

File: docs/rules/rules_of_hooks.md (hunk 3 above)
Comment (marvinhagemeister) [reply docs/rules/rules_of_hooks.md:73]: These are the only correct ones for all bad examples.

File: src/rules/rules_of_hooks.rs (hunk 5 above)
Comment (marvinhagemeister) [reply src/rules/rules_of_hooks.rs:277]: This is hit in pretty much all bad variants. It tests if the enclosing function is a component.

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**